src/main/java/com/github/manolo8/darkbot/gui/titlebar/ExtraButton.java

```diff
@@ -31,12 +35,58 @@
     private static final Set<ExtraMenuProvider> EXTRA_DECORATIONS = new LinkedHashSet<>();
     private static final Lazy<Void> clean = new Lazy.NoCache<>();
 
-    public static void setExtraDecorations(Stream<ExtraMenuProvider> provider) {
+    public static void setExtraDecorations(Stream<ExtraMenuProvider> provider, FeatureRegistry featureRegistry) {
         EXTRA_DECORATIONS.clear();
-        provider.forEach(EXTRA_DECORATIONS::add);
+        PluginExtraMenuProvider.PLUGINS.clear();
+
+        AtomicReference<ExtraMenuProvider> pluginProvider = new AtomicReference<>();
+        provider.forEach(extra -> {
+            if (extra instanceof PluginExtraMenuProvider) pluginProvider.set(extra);
+
+            Plugin pl = featureRegistry.getFeatureDefinition(extra).getPlugin();
+            if (pl != null && extra.autoSubmenu()) {
+                PluginExtraMenuProvider.PLUGINS.computeIfAbsent(
+                        pl,
+                        features -> new LinkedHashSet<>()).add(extra);
+            } else {
+                EXTRA_DECORATIONS.add(extra);
+            }
+        });
+
+        if (pluginProvider.get() != null) {
+            EXTRA_DECORATIONS.remove(pluginProvider.get());
+            EXTRA_DECORATIONS.add(pluginProvider.get());
+        }
+
         clean.send(null);
     }
 
+    @Feature(name = "Plugin menu provider", description = "Provides plugin extra menus")
+    public static class PluginExtraMenuProvider implements ExtraMenuProvider {
+        private static final Map<Plugin, Set<ExtraMenuProvider>> PLUGINS = new LinkedHashMap<>();
+
+        @Override
+        public Collection<JComponent> getExtraMenuItems(Main main) {
+            if (PLUGINS.isEmpty()) return Collections.emptyList();
+            List<JComponent> list = new ArrayList<>();
+
+            PLUGINS.forEach((plugin, features) -> {
+                features = features.stream()
+                        .filter(f -> !f.getExtraMenuItems(main).isEmpty())
+                        .collect(Collectors.toSet());
+                if (!features.isEmpty()) {
+                    list.add(createMenu(plugin.getName(),
+                            features.stream()
+                                    .flatMap(f -> f.getExtraMenuItems(main).stream())));
+                }
```

Comment:
You're calling getExtraMenuItems twice, and that *could* return two different things, creating an inconsistency. Also you're filtering features (which should remain non-modifed), and you're using `Collectors.toSet()` on them which kills ordering (it's not a linked hash set)

```suggestion
                List<JComponent> menus = features.stream()
                                    .flatMap(f -> f.getExtraMenuItems(main).stream()).collect(Collectors.toList());
                if (!menus.isEmpty())
                    list.add(createMenu(plugin.getName(), menus.stream());
```
instead of the final menus.stream() you can also add an overloaded method for createMenu that accepts any Iterable